### PR TITLE
Paginate sorting fix for SQL Server 2008

### PR DESCRIPTION
**Pull request by dexwerx**
https://github.com/bcosca/fatfree/issues/1039

Pagination fixes for SQL Server 2008
*limit/offset breaks when using subselect as a filter
*never had sorting+limit/offset working for SQL Server
*count() breaks when using a sort. SQL server doesnt like ORDER BY in a sub select, like the one used in mapper->count()

### CODE REVIEW

File: lib/db/sql/mapper.php [hunk 1]
```diff
@@ -248,30 +248,32 @@ function($str) use($db) {
...
-				if (!$options['order'])
-					$sql.=' ORDER BY '.$db->quotekey($pkeys[0]);
-				$sql.=' OFFSET '.$ofs.' ROWS';
+				$sql.=$ord.' OFFSET '.$ofs.' ROWS';
 				if ($lmt)
 					$sql.=' FETCH NEXT '.$lmt.' ROWS ONLY';
 			}
 			else {
 				// SQL Server 2008
-				$sql=str_replace('SELECT',
```
Comment (dexwerx) [inline lib/db/sql/mapper.php]: replacing all the SELECT's breaks subselects in the adhoc fields and filters. We really just want to replace the first one.

**Comment by dexwerx** [comment]
Would this better be created as a PR on fatfree-core?

**Comment by ikkez** [comment]
Yes, PR goes to fatfree-core.

I've used SQLServer 2008 in the past and didn't had any issues with sorting on paginated results, I havn't used sub-queries though - that probably needs a fix, but is the sorting issue a SQL Server 2005 only issue maybe? I can check your changes when I revived my old setup.
The `TOP 100 PERCENT` workaround looks a bit hackish to me. Isn't that something that should go into your subquery instead? btw: you doesn't check the server version at this point.. is that part harmless for all other sql server versions?

**Comment by dexwerx** [comment]
If you look at the code you can clearly see that the logic is hard coded to sort on the primary key when paginating, so I don't see how sorting ever worked with paging. Not only that but even without the subquery in the filter, the paginate call breaks in mapper->count() because SQL Server errors when trying to sort a sub select.

**Pull request closed without merging**